dfastmi/io/CelerObject.py

```diff
@@ -37,36 +37,52 @@
     CelerProperties
 
 """
-from abc import ABC, abstractmethod
-from typing import List
+from abc import ABC
+from typing import List, Tuple
+from pydantic import BaseModel
+from dfastmi.io.AReach import AReach
 
-
-class ICelerObject(ABC):
+class ICelerObject(ABC, BaseModel):
     "Interface or abstract base class to the CelerObject."
-    @abstractmethod
-    def validate(self):
-        pass
-    
-
-
+    parent_reach : AReach = None
+   
 class CelerDischarge(ICelerObject):
-    cdisch = tuple[float,float]
-    
-    def validate(self):
-        if self.cdisch == (0.0, 0.0):            
-            # raise Exception(
-            #             'The parameter "CelerQ" must be specified for branch "{}", reach "{}" since "CelerForm" is set to 2.'.format(
-            #                 branch,
-            #                 reach,
-            #             )
-            #         )
-            return
+    """
+        Series of discharge, bed celerity pairs
+    """
+    cdisch : Tuple[float,float] = (0.0, 0.0)
     
+    def validate(self, values):
```

Comment:
if you want to use automatic validation from Pydantic, you should decorate your functions with the `@validator` (v1) or `@model_validator` (v2) decorator.

V2 https://docs.pydantic.dev/latest/concepts/validators/#model-validators
V1 https://docs.pydantic.dev/1.10/usage/validators/

Reply:
With V2 you could write this as:

```
    @model_validator(mode='after')
    def validate(self) -> 'CelerDischarge':
        branch_name : str = self.parent_reach.parent_branch.name
        reach_name : str = self.parent_reach.name
        if self.cdisch == (0.0, 0.0):
             raise ValueError(f'The parameter "CelerQ" must be specified for branch "{branch_name}", '
                              f'reach "{reach_name}" since "CelerForm" is set to 2.')
        
        return self
```

The validation is automatically called directly after the initialization of the `CelerDischarge` object, and call to the super class is not needed, since validators are inherited by default.

Reply:
Thanks! really great... I thought i was doing it wrong... will update